File: leetcode/077-combinations/Combinations.java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

class Combinations {
    private void combine(List<List<Integer>> result, List<Integer> current, int start, int n, int k) {
        if (k == 0) {
            result.add(current.stream().map(Integer::new).collect(Collectors.toList()));
            return;
        }

        for (int i = start; i <= n; i++) {
            current.add(i);

            combine(result, current, i + 1, n, k - 1);

            current.remove(current.size() - 1);
        }
    }

    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList<>();

        combine(result, new ArrayList<>(), 1, n, k);

        return result;
    }
}

File: leetcode/077-combinations/combinations.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
const combine = function (n, k) {
  const nums = [...Array(n + 1).keys()];
  nums.shift();

  const result = [];
  const subResult = [];
  const backtrack = (start, k) => {
    if (k === 0) {
      result.push([...subResult]);
      return;
    }

    for (let i = start; i < nums.length; i++) {
      subResult.push(nums[i]);

      backtrack(i + 1, k - 1);

      subResult.pop();
    }
  };

  backtrack(0, k);

  return result;
};
